Convert kubeadm e2e Jenkins job to prow.

pipejakob opened the pull request:
After encountering a lot of problems getting this to work as a Jenkins job, I'm converting it to prow per EngProd's advice. This also allowed me to move the kubernetes-anywhere dependencies out of the Jenkins e2e image and into a new image specific to this job.

This is just the postsubmit CI job to start with. After testing, I'll add a presubmit job next.

CC @dmmcquay @mikedanese

dmmcquay commented:
@pipejakob this is awesome!

spxtr commented on images/ci-kubernetes-e2e-kubeadm/Dockerfile:
```diff
@@ -0,0 +1,63 @@
+ENV TERRAFORM_VERSION 0.7.2
+
+RUN apt-get update && \
+    apt-get install -y unzip
```
group with the others?

pipejakob replied:
Oh it's worse than that; this section was completely redundant given the other apt-get block (each section came from a different source Dockerfile I was borrowing from). I removed these entirely.

spxtr commented on jenkins/e2e-image/Dockerfile:
```diff
@@ -43,24 +43,6 @@ ENV FAIL_ON_GCP_RESOURCE_LEAK=true \
 # JENKINS_AWS_SSH_PUBLIC_KEY_FILE
 # JENKINS_AWS_CREDENTIALS_FILE
 
-# Add kubernetes-anywhere dependencies: jsonnet, terraform
```
Yay, this is awesome!

pipejakob commented:
For testing, I was going to accept that this would be an iterative process like adding the Jenkins job was in the first place, with some bugs only being found once the PR was merged/pushed.

What local testing I was able to do was:

  - Change the `PROJECT` var in the job script to one of my own personal projects
  - Comment out `git clone https://github.com/kubernetes/test-infra` in the runner so I could use my local repo with these unreleased changes
  - Run this command to approximate a job run:
```$ docker run -e REPO_NAME=kubernetes -e JOB_NAME=ci-kubernetes-e2e-kubeadm-gce -e GOOGLE_APPLICATION_CREDENTIALS=/etc/creds/creds.json -v $HOME/creds:/etc/creds/ -v $HOME/go/src/k8s.io/test-infra/:/workspace/test-infra -e BOOTSTRAP_MIGRATION=yes d827d65d42cd --branch master```

(where `d827d65d42cd` was the local id of the container built)

Is there additional local testing I could do to better simulate how this will actually run in prow? Last time we talked offline, I thought there wasn't.

pipejakob commented:
It's also worth mentioning that this job is dependent on https://github.com/kubernetes/test-infra/pull/1893. For local testing, I applied the change, rebuilt the kubekins-e2e image, and made the ci-kubernetes-e2e-kubeadm Dockerfile inherit from that rebuilt image.

Once https://github.com/kubernetes/test-infra/pull/1893 is merged and a new kubekins-e2e image is built, I'll need to update this Dockerfile to reference the new build (unless there's a better tag for it to use instead).

spxtr commented:
> Is there additional local testing I could do to better simulate how this will actually run in prow? Last time we talked offline, I thought there wasn't.

There isn't, but thanks for writing down this process. We should make an easier way. What you did should be sufficient for now.

> Once #1893 is merged and a new kubekins-e2e image is built, I'll need to update this Dockerfile to reference the new build (unless there's a better tag for it to use instead).

SG

/lgtm

pipejakob commented:
With https://github.com/kubernetes/test-infra/pull/1893 merged, I rebuilt and pushed the base e2e image, and updated this Dockerfile to refer to the new tag. Also, as a result of the Cloud SDK being installed in the e2e image, I've removed its section and a few dependencies from the Dockerfile as well.

Care to take another look?

spxtr commented on prow/postsubmit.yaml:
```diff
@@ -26,6 +26,16 @@ kubernetes/kubernetes:
         - "--clean"
         securityContext:
           privileged: true
+    run_after_success:
```
This is in `config.yaml` now.

spxtr commented:
Looks good after rebase.
